admin/services/restart.service.js

```diff
@@ -16,24 +24,91 @@ restartService.getPupils = async (schoolId) => {
   if (!school) throw new Error(`School [${schoolId}] not found`)
   let pupils = await pupilDataService.getSortedPupils(schoolId, 'lastName', 'asc')
-  pupils = pupils
-    .filter(p => restartService.isPupilEligible(p))
-    .map(({ _id, pin, dob, foreName, middleNames, lastName }) =>
+  pupils = await Promise.filter(pupils.map(async p => {
+    const isPupilEligible = await restartService.isPupilEligible(p)
+    if (isPupilEligible) return p
+  }), p => !!p)
+  if (pupils.length === 0) return []
+  pupils = pupils.map(({ _id, pin, dob, foreName, middleNames, lastName }) =>
       ({ _id, pin, dob: () => dateService.formatLongGdsDate(dob), foreName, middleNames, lastName }))
   pupils = pupilIdentificationFlagService.addIdentificationFlags(pupils)
   return pupils
 }
 
 /**
- * Determine if pupil is eligible for restart
+ * Determine all the factors that make pupil eligible for a restart
  * @param p
  * @returns {boolean}
  */
-restartService.isPupilEligible = (p) => {
-  // TODO: As part of the next PR pupil restart collection is introduced which will determine the number of restarts per pupil
-  if (p.attendanceCode) return false
+restartService.isPupilEligible = async (p) => {
+  const canRestart = await restartService.canRestart(p._id)
+  if (p.attendanceCode || !canRestart) return false
   const hasExpiredToday = p.pinExpiresAt && moment(p.pinExpiresAt).isSame(moment.now(), 'day')
   return !pinValidator.isActivePin(p.pin, p.pinExpiresAt) && hasExpiredToday
 }
 
+/**
+ * Perform restart for a list of pupils
+ * @param pupilsList
+ * @param restartReason
+ * @param didNotCompleteInfo
+ * @param restartFurtherInfo
+ * @param userId
+ * @returns {Promise.<void>}
+ */
+
+restartService.restart = async (pupilsList, restartReason, didNotCompleteInfo, restartFurtherInfo, userName) => {
+  await pinService.expireMultiplePins(pupilsList)
+  // All pupils should be eligible for restart before proceeding with creating a restart record for each one
+  const canAllPupilsRestart = restartService.canAllPupilsRestart(pupilsList)
+  if (!canAllPupilsRestart) {
+    throw new Error(`One of the pupils is not eligible for a restart`)
+  }
+  const submitted = await Promise.all(pupilsList.map(async pupilId => {
+    const pupilRestartData = {
+      pupilId,
+      recordedByUser: userName,
+      reason: restartReason,
+      didNotCompleteInformation: didNotCompleteInfo,
+      furtherInformation: restartFurtherInfo,
+      createdAt: moment.utc()
+    }
+    return pupilRestartDataService.create(pupilRestartData)
+  }))
+  return submitted
+}
+
+/**
+ * Checks if all selected pupils are eligible for restarts
+ * @param pupilsList
+ * @return {boolean}
+ */
+restartService.canAllPupilsRestart = async (pupilsList) => {
+  const eligibilityList = await Promise.all(pupilsList.map(async pupilId => {
+    const canRestart = await restartService.canRestart(pupilId)
+    if (!canRestart) return { pupilId, notEligible: true }
```

Comment:
I wonder if a the property name `notEligible` (for a restart) should actually be positive: e.g. calling it `eligible: false`, or simply {canRestart: <bool>}?

The `!e.notEligible` a few lines down is not particularly clear to me.